BLE_GattServer_FOTAService/scripts/test_fota.py

```diff
@@ -0,0 +1,268 @@
+# Copyright (c) 2020-2021 Embedded Planet
+# Copyright (c) 2009-2020 Arm Limited
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License
+
+import logging
+import asyncio
+import time
+
+from bleak          import BleakScanner, BleakClient, BleakError
+from bleak.uuids    import uuid16_dict
+from typing         import Optional
+from logging.config import fileConfig
+
+uuid16_dict = {v: k for k, v in uuid16_dict.items()}
+
+fileConfig('logging.conf')
+
+logger = logging.getLogger('logger')
+
+# FOTA UUIDS
+UUID_FOTA_SERVICE                  = "53880000-65fd-4651-ba8e-91527f06c887"
+UUID_BINARY_STREAM_CHAR            = "53880001-65fd-4651-ba8e-91527f06c887"
+UUID_CONTROL_CHAR                  = "53880002-65fd-4651-ba8e-91527f06c887"
+UUID_STATUS_CHAR                   = "53880003-65fd-4651-ba8e-91527f06c887"
+UUID_VERSION_CHAR                  = "53880004-65fd-4651-ba8e-91527f06c887"
+UUID_FIRMWARE_REVISION_STRING_CHAR = "0000{0:x}-0000-1000-8000-00805f9b34fb".format(
+    uuid16_dict.get("Firmware Revision String")
+)
+
+# FOTA status codes
+FOTA_STATUS_OK                     = bytearray(b'\x00')
+FOTA_STATUS_UPDATE_SUCCESSFUL      = bytearray(b'\x01')
+FOTA_STATUS_XOFF                   = bytearray(b'\x02')
+FOTA_STATUS_XON                    = bytearray(b'\x03')
+FOTA_STATUS_SYNC_LOST              = bytearray(b'\x04')
+FOTA_STATUS_UNSPECIFIED_ERROR      = bytearray(b'\x05')
+FOTA_STATUS_VALIDATION_FAILURE     = bytearray(b'\x06')
+FOTA_STATUS_INSTALLATION_FAILURE   = bytearray(b'\x07')
+FOTA_STATUS_OUT_OF_MEMORY          = bytearray(b'\x08')
+FOTA_STATUS_MEMORY_ERROR           = bytearray(b'\x09')
+FOTA_STATUS_HARDWARE_ERROR         = bytearray(b'\x0a')
+FOTA_STATUS_NO_FOTA_SESSION        = bytearray(b'\x0b')
+
+# FOTA op codes
+FOTA_OP_CODE_START                 = bytearray(b'\x01')
+FOTA_OP_CODE_STOP                  = bytearray(b'\x02')
+FOTA_OP_CODE_COMMIT                = bytearray(b'\x03')
+FOTA_OP_CODE_SET_XOFF              = bytearray(b'\x41')
+FOTA_OP_CODE_SET_XON               = bytearray(b'\x42')
+FOTA_OP_CODE_SET_FRAGMENT_ID       = bytearray(b'\x43')
+
+# Number of bytes in a single BSC packet, excluding fragment ID
+FRAGMENT_SIZE       = 128
```

Comment:
It can't be 19 as that is not a multiple of the program block size (= 4 for nRF & DISCO). In general, this is a QSPIF configuration option that we can override in `mbed_app.json`:

``` JSON
{
    "target_overrides": {
        "*": {
            "QSPI_MIN_READ_SIZE": "19",
            "QSPI_MIN_PROG_SIZE": "19"
        }
    }
}
```
Since 19 is prime. That will only work for nRF and DISCO which use QSPIF. Targets with a different default block device instance will require separate overrides.

Reply:
That's interesting, we probably need some sort of buffering in the client side to accommodate between the block size and packet length.

Reply:
I'll leave this open as it's a good observation and the bleak PR is worth keeping an eye. However, a fragment size of 128 data bytes + 1 fragment ID byte (=129 bytes) should work on nRF and DISCO due to the `cordio.desired-att-mtu` override in our `mbed-app.json`. Making this 19 would also affect the transfer speed. We can iterate on this PR once it lands in the FOTA service example branch. For now, perhaps we can simply add a note about hardware support in the README?